{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Script para construir el fenotipo en formato tensorqtl: con una sola lÃ­nea de encabezado que comience con # y las primeras cuatro columnas correspondientes a: chr, start, end, phenotype_id, con las columnas restantes correspondientes a las muestras (los identificadores deben coincidir con los de la entrada del genotipo). La salida es un archivo con formato bed. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from collections import defaultdict\n",
    "import subprocess\n",
    "import gzip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def to_bgzip(df, path, header=True, float_format=None):\n",
    "    \"\"\"Write DataFrame to bgzip\"\"\"\n",
    "    assert path.endswith('.gz')\n",
    "    bgzip = subprocess.Popen('bgzip -c > '+path, stdin=subprocess.PIPE, shell=True, encoding='utf8')\n",
    "    df.to_csv(bgzip.stdin, sep='\\t', index=False, header=header, float_format=float_format)\n",
    "    stdout, stderr = bgzip.communicate()\n",
    "    subprocess.check_call('tabix -f '+path, shell=True)\n",
    "\n",
    "\n",
    "def sort_bed(bed_df, inplace=True):\n",
    "    \"\"\"Sort BED DataFrame\"\"\"\n",
    "    sorted_df = bed_df.sort_values(['chr', 'start', 'end'], key=lambda x:\n",
    "                    x.str.replace('chr','').str.replace('X','23').astype(int) if x.dtype == object else x,\n",
    "                    inplace=inplace)\n",
    "    if inplace:\n",
    "        bed_df.reset_index(drop=True, inplace=True)\n",
    "    else:\n",
    "        sorted_df.reset_index(drop=True, inplace=True)\n",
    "        return sorted_df\n",
    "\n",
    "\n",
    "def write_bed(bed_df, output_name, header=True, float_format=None):\n",
    "    \"\"\"Write DataFrame to BED format\"\"\"\n",
    "    if header:  \n",
    "        assert (bed_df.columns[0]=='chr' or bed_df.columns[0]=='#chr') and bed_df.columns[1]=='start' and bed_df.columns[2]=='end'\n",
    "        # header must be commented in BED format\n",
    "        header = bed_df.columns.values.copy()\n",
    "        header[0] = '#chr'\n",
    "    to_bgzip(bed_df, output_name, header=header, float_format=float_format)\n",
    "\n",
    "\n",
    "def read_gct(gct_file, sample_ids=None, dtype=None, load_description=True, skiprows=2):\n",
    "    \"\"\"Load GCT as DataFrame\"\"\"\n",
    "    if sample_ids is not None:\n",
    "        sample_ids = ['Name']+list(sample_ids)\n",
    "\n",
    "    if gct_file.endswith('.gct.gz') or gct_file.endswith('.gct'):\n",
    "        if dtype is not None:\n",
    "            with gzip.open(gct_file, 'rt') as gct:\n",
    "                for _ in range(skiprows):\n",
    "                    gct.readline()\n",
    "                sample_ids = gct.readline().strip().split()\n",
    "            dtypes = {i:dtype for i in sample_ids[2:]}\n",
    "            dtypes['Name'] = str\n",
    "            dtypes['Description'] = str\n",
    "            df = pd.read_csv(gct_file, sep='\\t', skiprows=skiprows, usecols=sample_ids, index_col=0, dtype=dtypes)\n",
    "        else:\n",
    "            df = pd.read_csv(gct_file, sep='\\t', skiprows=skiprows, usecols=sample_ids, index_col=0)\n",
    "    elif gct_file.endswith('.parquet'):\n",
    "        df = pd.read_parquet(gct_file, columns=sample_ids)\n",
    "    else:\n",
    "        raise ValueError('Unsupported input format.')\n",
    "    if not load_description and 'Description' in df.columns:\n",
    "        df.drop('Description', axis=1, inplace=True)\n",
    "    return df\n",
    "\n",
    "\n",
    "def write_gct(df, gct_file, float_format='%.6g'):\n",
    "    \"\"\"Write DataFrame to GCT format\"\"\"\n",
    "    assert df.index.name=='Name' and df.columns[0]=='Description'\n",
    "    if gct_file.endswith('.gct.gz'):\n",
    "        opener = gzip.open(gct_file, 'wt', compresslevel=6)\n",
    "    else:\n",
    "        opener = open(gct_file, 'w')\n",
    "\n",
    "    with opener as gct:\n",
    "        gct.write(f'#1.2\\n{df.shape[0]:d}\\t{df.shape[1]-1:d}\\n')\n",
    "        df.to_csv(gct, sep='\\t', float_format=float_format)\n",
    "\n",
    "\n",
    "def gtf_to_tss_bed(annotation_gtf, feature='gene', exclude_chrs=[], phenotype_id='gene_id'):\n",
    "    \"\"\"Parse genes and TSSs from GTF and return DataFrame for BED output\"\"\"\n",
    "    chrom = []\n",
    "    start = []\n",
    "    end = []\n",
    "    gene_id = []\n",
    "    gene_name = []\n",
    "\n",
    "    if annotation_gtf.endswith('.gz'):\n",
    "        opener = gzip.open(annotation_gtf, 'rt')\n",
    "    else:\n",
    "        opener = open(annotation_gtf, 'r')\n",
    "\n",
    "    with opener as gtf:\n",
    "        for row in gtf:\n",
    "            row = row.strip().split('\\t')\n",
    "            if row[0][0]=='#' or row[2]!=feature: continue # skip header\n",
    "            chrom.append(row[0])\n",
    "\n",
    "            # TSS: gene start (0-based coordinates for BED)\n",
    "            if row[6]=='+':\n",
    "                start.append(np.int64(row[3])-1)\n",
    "                end.append(np.int64(row[3]))\n",
    "            elif row[6]=='-':\n",
    "                start.append(np.int64(row[4])-1)  # last base of gene\n",
    "                end.append(np.int64(row[4]))\n",
    "            else:\n",
    "                raise ValueError('Strand not specified.')\n",
    "\n",
    "            attributes = defaultdict()\n",
    "            for a in row[8].replace('\"', '').split(';')[:-1]:\n",
    "                kv = a.strip().split(' ')\n",
    "                if kv[0]!='tag':\n",
    "                    attributes[kv[0]] = kv[1]\n",
    "                else:\n",
    "                    attributes.setdefault('tags', []).append(kv[1])\n",
    "\n",
    "            gene_id.append(attributes['gene_id'])\n",
    "            gene_name.append(attributes['gene_name'])\n",
    "\n",
    "    if phenotype_id=='gene_id':\n",
    "        bed_df = pd.DataFrame(data={'chr':chrom, 'start':start, 'end':end, 'gene_id':gene_id}, columns=['chr', 'start', 'end', 'gene_id'], index=gene_id)\n",
    "    elif phenotype_id=='gene_name':\n",
    "        bed_df = pd.DataFrame(data={'chr':chrom, 'start':start, 'end':end, 'gene_id':gene_name}, columns=['chr', 'start', 'end', 'gene_id'], index=gene_name)\n",
    "    # drop rows corresponding to excluded chromosomes\n",
    "    mask = np.ones(len(chrom), dtype=bool)\n",
    "    for k in exclude_chrs:\n",
    "        mask = mask & (bed_df['chr']!=k)\n",
    "    bed_df = bed_df[mask]\n",
    "\n",
    "    # sort by start position\n",
    "    bed_df = bed_df.groupby('chr', sort=False, group_keys=False).apply(lambda x: x.sort_values('start'))\n",
    "\n",
    "    return bed_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
